libraries/chain/protocol/fee_schedule.cpp

```diff
@@ -115,21 +115,23 @@ namespace graphene { namespace chain {
       this->scale = 0;
    }
 
+   asset fee_schedule::calculate_fee( const operation& op )const
+   {
+      uint64_t required_fee = op.visit( calc_fee_visitor( *this, op ) );
+      if( scale != GRAPHENE_100_PERCENT )
+      {
+         auto scaled = fc::uint128(required_fee) * scale;
```

Comment:
Possible optimization: if `required_fee <= UINT32_MAX` then can use 64 bit multiplication (`scale` is `uint32_t`). Add static_assert for type of `scale` to be safe.

Reply:
Actually I didn't think much about how to optimize the code inside the `if`, because IMHO `scale` would be always identical to `GRAPHENE_100_PERCENT` in the future, it was a poor decision made in the past (I don't have the link to the discussion right now, if need I'll try to find it).

Reply:
Here is a discussion related to `scale`: https://bitsharestalk.org/index.php?topic=20073.0 . Although perhaps there is no bug, people did get confused when `scale` is not `1`.

BTW `scale` was introduced due to this discussion: https://github.com/cryptonomex/graphene/issues/66

Reply:
Uh, 8 pages?! tl;dr please?

Reply:
The story in 2015:
* Initial transfer fee of BitShares mainnet was 20 BTS. Some time after launched, BM updated `scale` to `2` when he was directly controlling the committee (via initX accounts), so transfer fee became 40 BTS (`transfer_fee = 20` * `scale = 2`);
* a bug was found which caused referrers not receiving most of referral income, in BM's reply he used the word "scale";
* new committee members found the `scale` parameter in fee schedule which is not `1`, thought it caused the bug, so decided to reset `scale` to `1` and set the fees to intuitive values. Actually the bug is not related to the parameter.
* After that, the `scale` parameter never got changed, stayed at `1` forever.

About the parameter:
* It's introduced due to the use case "all the prices could be defined in USD" so the Committee is able to update only one parameter to effectively update all fees,
  * the first change of `scale` (from 1 to 2) was not due to this use case,
  * there was no discussion about this use case when the Committee was resetting it to 1,
  * current fee schedule is USD dominated, but scale was always 1.
* During the discussion some people were confused, because https://cryptofresh.com shows a scaled fee schedule but not the raw data.
* IMHO, it's better to be intuitive, keep `scale` at `1` and define individual fees in actual values.

By the way, when I was developing BSIP10 (percentage-based transfer fee), I found the `scale` parameter should be ignored when transferring assets with percentage-based fee configured. E.G. if the percentage is a 1% and `scale` is 100, it's not expected to pay a 100% fee.

Reply:
It doesn't harm to optimize the code though. I'm just not in the mood for it right now ;)

Reply:
Thanks. Seems like a low priority optimization then.

Reply:
Note: `boost::multiprecision::cpp_int` knows how to quickly multiply small numbers. So I'd like to do this optimization in the scope of #1660.